public/.htaccess

```diff
@@ -21,6 +21,7 @@
 </IfModule>
 
 Options -Indexes
+SetEnv no-gzip
```

Comment:
It is required to reliably enable streaming responses. Without this option Apache buffers the whole answer generated by PHP in order to compress it and then transfer it with the header `Encoding: gzip` or `Encoding: deflate`. This buffering prevents streaming responses from working and we need them for the progress and status report during import from server.

However, Apache uses compression only under very specific circumstances:

1. The module `mod_gzip` or `mod_deflate` must be installed and enabled
2. You don't use `mod_php`, but `mod_proxy_fcgi` together with PHP FPM.
3. The browser (and any potential content delivery proxy in between) advertises that it supports compressed connections

If all three conditions apply _simultaneously_, then and only then gzip or deflate kicks in. In that case you need to set `no_gzip` in order to still ensure that streaming responses work as expected. It took me nearly two days to find that out while I was working on that part of the backend. I was unable to get a streamed response, but the whole response came at once in the end.

The answer was deeply buried in the official Apache docs and I don't find the link anymore. Currently, I only find this obscure website: https://www.jeffgeerling.com/blog/2016/streaming-php-disabling-output-buffering-php-apache-nginx-and-varnish. Search for the phrase "If you use mod_proxy_fcgi with PHP-FPM"

I believe nobody noticed this problem before, because point 2 is a rather rare setup. I guess most user simply use `mod_php`, although it is rather slow, legacy and prevents Apache from multi-threading. Even vanilla Apache does not use `mod_php` anymore and recommend against it, but most of the big distros (Debian, Ubuntu, ...) still use it as default. I guess I am the only crazy person who uses `mod_proxy_fcgi` + PHP FPM. It makes PHP so much faster :sunglasses:

Moreover, neither `mod_gzip` nor `mod_deflate` are default Apache modules in Debian or Ubuntu. My productive system at my web hoster is Ubuntu-based, so I never had run into the problem before. However, my development machine at home uses Gentoo and Gentoo installs `mod_gzip` by default. Et voila, ..., the problem occurred.

I guess, it always should have been set, but nobody notices so far.

Of course, we could set the option `no_gzip` more locally via a fancy Apache rewrite rule only for the URLs that need it instead of disabling compression for Lychee on a global level. But I decided that is is not worth the effort. The real content of Lychee are images and these files are already highly compressed, so another layer of compression would rather be harmful. The remaining responses are short JSON snippets with a few bytes, so the total number of saved bytes is small, too.

Reply:
Understood. I have compression enabled in Apache but I use `mod_php`... I can imagine that responses to calls like `Album::get` can get rather large for big albums so they could benefit from compression, but if it's nontrivial to disable compression just for the streaming output, it's probably not worth it...

Reply:
> I can imagine that responses to calls like Album::get can get rather large for big albums so they could benefit from compression,

Yes. This also has been the only response, I could think of, whose size is not constant but grows linearly (and the log listing, of course). But even if an album contains 1,000 photos, a photo entry has around 200 characters (I didn't count them, just a gut feeling). So we end up with 200kB. 


> but if it's nontrivial to disable compression just for the streaming output, it's probably not worth it...

I actually did not spend much time to even try it. I was just happy after I had finally managed to get streamed responses to work. It was really frustrating. Honestly, I find the syntax for Apache's rewrite rules extremely unintuitive. It feels like the 70s with all the single letter commands. While O really like Apache a lot better than Nginx, the rewrite system is just, meeeeeh

Either we try to fix it now or in a separate PR.

Reply:
A more fine-granular settings was actually really easy. Fixed it in https://github.com/LycheeOrg/Lychee/pull/1217/commits/1cc7bc74062f8c58422654944e02cb442916a913